### PR TITLE
Fix quantity rules when variant is switched

### PR DESCRIPTION
**Pull request by KaichenWang**
## PR Summary

Fixes issue with quantity rules not updating correctly on product page and featured product section when variant is switched.

## Why are these changes introduced?

On Product page and Featured product section, switching to another variant with different quantity rules does not update quantity rules messaging correctly.

#### Expected behavior
https://user-images.githubusercontent.com/3710548/214176284-995059b0-54c1-4d8e-9dfa-543cbcb7e4aa.mp4

## What approach did you take?

- Add `id` to the quantity input form in order to target the element in JS
- Add JS to function which is triggered when variant is switched and new markup is returned async
- Use `id` to fetch current element on page and updated element in new markup
- Replace the current element in the DOM with updated element

## Testing steps/scenarios

Ping me for detailed tophat instructions (requires B2B account)
- Navigate to product page
- [ ] Verify that different variants with different quantity rules have rules updated after switching variants
- Navigate to home page and find featured product section
- [ ] Verify that different variants with different quantity rules have rules updated after switching variants

## Demo links

- [Store](https://kaibuyz.myshopify.com/?preview_theme_id=139287330838)
- [Editor](https://kaibuyz.myshopify.com/admin/themes/139287330838/editor)

## Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/global.js [hunk 1]
```diff
@@ -906,6 +906,7 @@ class VariantSelects extends HTMLElement {
 
   renderProductInfo() {
     const requestedVariantId = this.currentVariant.id;
+    const sectionId = this.dataset.originalSection ? this.dataset.originalSection : this.dataset.section;
```
Comment (KaichenWang) [inline assets/global.js:909]: I'm not sure what `this.dataset.originalSection` is referring to, but I'm keeping it to be safe. Maybe @LucasLacerdaUX can provide some context since it was added in this [PR](https://github.com/Shopify/dawn/pull/1388/files).

File: assets/product-info.js [hunk 2]
```diff
@@ -54,21 +53,16 @@ if (!customElements.get('product-info')) {
...
-            this.destinationQty.innerHTML = valueQtyCart;
-          }
-        } else {
-          if (this.input) this.input.dataset.cartQuantity = 0;
-          this.destinationQty.innerHTML = 0;
-          this.querySelector(quantityRulesCartClassname).classList.add('hidden')
+        const html = new DOMParser().parseFromString(responseText, 'text/html');
+        const source = html.querySelector((`[data-quantity-variant-id~="${this.currentVariant.value}"]`));
+        
+        if (source) {
```
Comment (KaichenWang) [inline assets/product-info.js:59]: - Simplified the logic
- Use `querySelector()` to get updated element instead of using cached element, because variant change now replaces the input in the DOM

**Comment by bakura10** [comment]
I was thinking of doing something similar in our themes, however I always had two concerns:

* isn’t changing the quantity without customer consent not also a bit confusing ? I have selected 5, and the theme changes automatically my quantity. I may not realised it has changed and add a quantity I did not expect.
* in the past inventory info was removed from the product | json to avoid leaking inventory. I feel this change actually make the inventory kind of visible again through the quantity selector.

**Review by sofiamatulis — COMMENTED** [review]
Nice changes 👍 Added a few comments

File: sections/featured-product.liquid [hunk 3]
```diff
@@ -190,6 +190,7 @@
                 </p>
               {%- when 'quantity_selector' -%}
                 <div
+                  id="Quantity-Form-{{ section.id }}"
                   class="product-form__input product-form__quantity{% if settings.inputs_shadow_vertical_offset != 0 and settings.inputs_shadow_vertical_offset < 0 %} product-form__quantity-top{% endif %}"
                   {{ block.shopify_attributes }}
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply sections/featured-product.liquid:195]: Good find! Just checked and this is an existing bug in Dawn `7.0.1` (see video below), I can open an issue for this. I think this wasn't caught because opening the cart drawer on the cart page is an edge case and not often tested.

https://user-images.githubusercontent.com/3710548/214688931-7790080e-4514-4923-9c3c-3cca7daac38b.mp4

[the remark this answers is not in the document]
Comment (sofiamatulis) [reply sections/featured-product.liquid:195]: Thanks for checking! I agree this can definitely be tackled separately plus it is an edge case a well 👍

File: assets/product-info.js [hunk 4]
```diff
@@ -54,21 +53,16 @@ if (!customElements.get('product-info')) {
...
+          const quantityInCart = parseInt(source.value);
+          this.querySelector('.quantity__input').dataset.cartQuantity = quantityInCart;
+          this.querySelector('.quantity-cart').innerHTML = quantityInCart;
+          this.querySelector('.quantity__rules-cart').classList.toggle('hidden', quantityInCart <= 0);
         }
+
         this.fetchQtyRules();
       })
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/product-info.js:67]: 1. We're not caching the unsubmitted quantity value. When we switch variants, the quantity will be reset to the minimum `{{product.selected_or_first_available_variant.quantity_rule.min }}` value from Liquid for that variant. So going from `63` to `49` is expected. Although we could improve this by caching the last entered value, I think as it is currently, it's not difficult to understand from a buyer perspective.

2. The cart quantity can affect the variant's actual minimum. This calculation is done in JS in the `setQuantityBoundries()` method, but only after the quantity element is appended to the DOM (this DOM update happens on variant change). This results in the flash from `49` to `7`. I agree this is not ideal, but requires some refactoring in our logic (I can open an issue). I would label this as a UX enhancement, so i don't think it should block this PR.

File: assets/product-info.js [hunk 5]
```diff
@@ -54,21 +53,16 @@ if (!customElements.get('product-info')) {
...
-          this.querySelector(quantityRulesCartClassname).classList.toggle('hidden', valueQtyCart <= 0);
-          if (valueQtyCart && this.input) {
-            this.input.dataset.cartQuantity = valueQtyCart;
-            this.destinationQty.innerHTML = valueQtyCart;
-          }
-        } else {
-          if (this.input) this.input.dataset.cartQuantity = 0;
-          this.destinationQty.innerHTML = 0;
-          this.querySelector(quantityRulesCartClassname).classList.add('hidden')
+        const html = new DOMParser().parseFromString(responseText, 'text/html');
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/product-info.js]: Yes, good catch. Fixed - https://github.com/Shopify/dawn/pull/2251/commits/8556d78db6f42df6a2c98ba74e87bf805cafa150

File: assets/global.js [hunk 6]
```diff
@@ -935,6 +936,11 @@ class VariantSelects extends HTMLElement {
         if (inventoryDestination) inventoryDestination.classList.toggle('visibility-hidden', inventorySource.innerText === '');
 
         this.toggleAddButton(!this.currentVariant.available, window.variantStrings.soldOut);
+
+        // Update quantity form
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js]: Removed - https://github.com/Shopify/dawn/pull/2251/commits/f1a17699fee4af1c854804c259cb0471099bc42b

File: assets/global.js [hunk 7]
```diff
@@ -935,6 +936,11 @@ class VariantSelects extends HTMLElement {
...
+        const quantityFormCurrent = document.getElementById(`Quantity-Form-${sectionId}`);
+        const quantityFormUpdated = html.getElementById(`Quantity-Form-${sectionId}`);
+        if (quantityFormCurrent && quantityFormUpdated) quantityFormCurrent.innerHTML = quantityFormUpdated.innerHTML;
       });
   }
```
[the remark this answers is not in the document]
Comment (sofiamatulis) [reply assets/global.js:944]: I agree we shouldnt show it at the bottom because the qty rules change depending on the variant, but the information is still important. What about in here?

https://screenshot.click/26-49-mmpox-2ng3f.png

Screenshot 2023-01-26 at 9 50 40 AM

**Review by sofiamatulis — COMMENTED** [review]
Looks good 🎉 Thanks for the changes

I just have one small suggestion for no JS - let me know your thoughts

https://github.com/Shopify/dawn/pull/2251#discussion_r1087951451

File: assets/global.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js:944]: Good idea! Updated - https://github.com/Shopify/dawn/pull/2251/commits/76ba1b30123cdadb4cf9168c35119a708a89e025

[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js:944]: Also noticed that [this commit](https://github.com/Shopify/dawn/commit/a422c45be43027628f02c77f8ae98f7c6d34b88c) broke no-js add-to-cart. Fixed this issue - https://github.com/Shopify/dawn/pull/2251/commits/e18fb6131a61d9507e0750e4acacf076a73e02e9

**Comment by bakura10** [comment]
Do you have documentation about this feature?

**Review by sofiamatulis — APPROVED** [review]
Looks good! 🚢

**Review by edmund-teh — DISMISSED** [review]
Working as expected in a variety of test scenarios for me... thanks @KaichenWang!

File: assets/global.js [hunk 8]
```diff
@@ -935,6 +936,10 @@ class VariantSelects extends HTMLElement {
         if (inventoryDestination) inventoryDestination.classList.toggle('visibility-hidden', inventorySource.innerText === '');
 
         this.toggleAddButton(!this.currentVariant.available, window.variantStrings.soldOut);
+
+        const quantityFormCurrent = document.getElementById(`Quantity-Form-${sectionId}`);
+        const quantityFormUpdated = html.getElementById(`Quantity-Form-${sectionId}`);
+        if (quantityFormCurrent && quantityFormUpdated) quantityFormCurrent.innerHTML = quantityFormUpdated.innerHTML;
```
Comment (LucasLacerdaUX) [inline assets/global.js:942]: Is there any alternative to replacing this element with a new one, such as updating values to match? 

If possible, I would love to avoid replacing elements because, especially on slower connections, users can quickly tab to the input and start working on it while the request is still being processed. Once the request is completed, the active element is replaced and, as a result, keyboard focus is lost.

https://user-images.githubusercontent.com/37168033/215180779-42c51357-948a-4cc4-a5b1-035cea0a019d.mp4

Ideally, we would replace only the values that are needed (max, min, value, etc) via JS and not completely replace the element.

Alternatively, we could block user interactions with the quantity inputs until these elements are updated.

**Review by LucasLacerdaUX — COMMENTED** [review]
Some quick questions. I'll continue the review and leave comments individually as I find them.

File: snippets/product-variant-picker.liquid [hunk 9]
```diff
@@ -86,9 +90,29 @@
...
+              if variant.quantity_rule.increment > 1
+                echo 'products.product.quantity.multiples_of' | t: quantity: variant.quantity_rule.increment | prepend: ' - '
+              endif
+              if variant.quantity_rule.min > 1
+                echo 'products.product.quantity.minimum_of' | t: quantity: variant.quantity_rule.min | prepend: ' - '
+              endif
+              if variant.quantity_rule.max != null
+                echo 'products.product.quantity.maximum_of' | t: quantity: variant.quantity_rule.max | prepend: ' - '
+              endif
+              # TODO: enable theme-check once `item_count_for_variant` is accepted as valid filter
```
[the remark this answers is not in the document]
Comment (LucasLacerdaUX) [reply snippets/product-variant-picker.liquid:108]: Sounds good! Let's keep it for now and remove if it's ready by the time we create the releases. It really won't impact the final zip anyway.

File: assets/cart.js [hunk 10]
```diff
@@ -121,7 +122,10 @@ class CartItems extends HTMLElement {
             this.getSectionInnerHTML(parsedState.sections[section.section], section.selector);
         }));
         const updatedValue = parsedState.items[line - 1].quantity;
-        const message = updatedValue !== parseInt(quantityElement.value) ? window.cartStrings.quantityError.replace('[quantity]', updatedValue) : '';
+        let message = '';
+        if (items.length === parsedState.items.length && updatedValue !== parseInt(quantityElement.value)) {
```
[the remark this answers is not in the document]
Comment (sofiamatulis) [reply assets/cart.js:126]: Can you share more context on this bug?

**Comment by LucasLacerdaUX** [comment]
Just something I noticed while playing with the cart drawer. I'm still investigating why this is happening in the first place, but wanted to share this early so you can jump if you have any idea on why this is happening. This happens on a non-B2B scenario.

image

Steps to reproduce:
- Add two different variants of the same item to the cart
- Remove one of them
- There'll be a `There was an error while updating your cart. Please try again.`

Removing items from the cart drawer also won't update the quantity label, while editing quantity on drawer does.

Here's a quick video:
https://user-images.githubusercontent.com/37168033/215575730-b63e40da-ece7-4a87-bde3-b38f45776b1b.mp4

File: assets/cart.js [hunk 11]
```diff
@@ -100,6 +100,7 @@ class CartItems extends HTMLElement {
       .then((state) => {
```
Comment (sofiamatulis) [inline assets/cart.js:100]: When you first get to the page it shows as `sold out` and gets fixed once you switch variants:

https://screenshot.click/31-33-ef46b-6iy9h.mp4

File: assets/cart.js [hunk 12]
```diff
@@ -100,6 +100,7 @@ class CartItems extends HTMLElement {
...
         const parsedState = JSON.parse(state);
         const quantityElement = document.getElementById(`Quantity-${line}`) || document.getElementById(`Drawer-quantity-${line}`);
+        const items = document.querySelectorAll('.cart-item');
 
         if (parsedState.errors) {
```
Comment (sofiamatulis) [inline assets/cart.js:105]: The same way the error of min shows, shouldnt the error for max show as well? 🤔 

https://screenshot.click/31-39-0dbqc-vsfu2.mp4

Comment (sofiamatulis) [reply assets/cart.js:105]: Also after it changes to 84, the +/- dont seem to work anymore and keep defaulting back to 84

https://screenshot.click/31-41-o9daq-g2q1q.mp4

File: assets/base.css [hunk 13]
```diff
@@ -1786,7 +1786,7 @@ details[open] > .share-button__fallback {
 .customer select {
```
Comment (sofiamatulis) [inline assets/base.css:1786]: I activated the buy 2, get 1 50% off and I am not seeing the discount:

https://screenshot.click/31-14-n2gfl-s4h4r.png

(under discounts in admin)

File: assets/cart.js (hunk 11 above)
Comment (KaichenWang) [reply assets/cart.js:100]: Good catch, the Liquid logic for disabling add-to-cart button was [updated](https://github.com/Shopify/dawn/pull/2255), but our logic for variant switching wasn't. Updated - https://github.com/Shopify/dawn/pull/2251/commits/a3cc6f65f1db654ac2d1206c12a207f44c56295a

**Comment by LucasLacerdaUX** [comment]
I'm starting a new round of tests now. Hoping to finish review by EOD.

I've been testing for pretty much the whole day on both B2B and non-B2B and wasn't able to find new bugs other than the Add to Cart one Sofia mentioned above.

File: assets/cart.js (hunk 12 above)
Comment (sofiamatulis) [reply assets/cart.js:105]: Cart drawer was enabled good point 👍 

What about the min error? 👀 https://github.com/Shopify/dawn/pull/2251#discussion_r1092114873

File: assets/product-info.js [hunk 14]
```diff
@@ -45,57 +48,43 @@ if (!customElements.get('product-info')) {
...
+        this.querySelector('.quantity__rules-cart .loading-overlay').classList.add('hidden');
       });
     }
+
+    updateQuantityRules(sectionId, html) {
+      const quantityFormCurrent = document.getElementById(`Quantity-Form-${sectionId}`);
+      const quantityFormUpdated = html.getElementById(`Quantity-Form-${sectionId}`);
+      const selectors = ['.quantity__input', '.quantity__rules', '.quantity__label'];
+      for (let selector of selectors) { 
+        const current = quantityFormCurrent.querySelector(selector);
```
Comment (LucasLacerdaUX) [inline assets/product-info.js:74]: Variant switching is broken inside Quick Buy because of this line. Another error, on the same line, appears when you add an item to your cart.

We use `Quantity-Form-quickadd-template--17128870936598__main` for Quick Buy, while `Quantity-Form-template--17128870936598__main` is used outside. That's why we had the `data-original-section` and `data-section` in the variant switching components.

We have to change ids for Quick Add to avoid duplicates (ex.: if you open a quick Modal for **Product 1** on while you're viewing that same product page).

Comment (KaichenWang) [reply assets/product-info.js]: Thanks for double-checking! I admittedly had not tested the quick add flow properly. Updated - https://github.com/Shopify/dawn/pull/2251/commits/1995689ba775878e6e98f95f628640a1e8f9df57

File: assets/product-info.js [hunk 15]
```diff
@@ -6,24 +6,27 @@ if (!customElements.get('product-info')) {
...
+    variantChangeUnsubscriber = undefined;
 
     connectedCallback() {
       this.setQuantityBoundries();    
-      this.cartUpdateUnsubscriber = subscribe(PUB_SUB_EVENTS.cartUpdate, this.fetchCartQty.bind(this));
+      this.cartUpdateUnsubscriber = subscribe(PUB_SUB_EVENTS.cartUpdate, this.fetchQuantityRules.bind(this));
+      this.variantChangeUnsubscriber = subscribe(PUB_SUB_EVENTS.variantChange, (event) => {
+        this.updateQuantityRules(event.data.sectionId, event.data.html);
+        this.setQuantityBoundries();
+      });
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/product-info.js]: Good point! Updated - https://github.com/Shopify/dawn/pull/2251/commits/18272fc156595b75035c944ad1a4d15a577d4f75

File: assets/global.js [hunk 16]
```diff
@@ -934,7 +935,13 @@ class VariantSelects extends HTMLElement {
...
         if (inventoryDestination) inventoryDestination.classList.toggle('visibility-hidden', inventorySource.innerText === '');
 
-        this.toggleAddButton(!this.currentVariant.available, window.variantStrings.soldOut);
+        const addButtonUpdated = html.getElementById(`ProductSubmitButton-${sectionId}`);
+        this.toggleAddButton(addButtonUpdated ? addButtonUpdated.hasAttribute('disabled') : true, window.variantStrings.soldOut);
+
+        publish(PUB_SUB_EVENTS.variantChange, {data: {
+          sectionId,
+          html
+        }});
```
Comment (gideonb) [inline assets/global.js:944]: I think if you add something called `variantChange` to a global pub/sub, a lot of folks in the ecosystem will subscribe to it expecting useful information that will inform their subscription apps, bundle popups, custom carts, sizing widgets, and so on. They are going to be super disappointed when there's no `data.variant` inside.

Comment (KaichenWang) [reply assets/global.js]: Good point. Updated - https://github.com/Shopify/dawn/pull/2251/commits/c668a7beac728fd6a5438546c678610a670760a7

**Review by sofiamatulis — COMMENTED** [review]
Everything seems to be working well! I just dont see complimentary products when I log in with b2b 🤔 

|   | Cart Drawe | Cart Popup | Cart Page | Complimentary Products | Quick Add | Variant Picker Pill | Variant Picker Dropdown | no JS | PDP | Feat product |
|---|------------|------------|-----------|------------------------|-----------|---------------------|-------------------------|-------|-----|--------------|
|   | ✅          | ✅          | ✅         | ❌                      | ✅         | ✅                   | ✅                       | ✅     | ✅   | ✅            |
|   |            |            |           |                        |           |                     |                         |       |     |              |
|   |            |            |           |                        |           |                     |                         |       |     |              |
|   |            |            |           |                        |           |                     |                         |       |     |              |
|   |            |            |           |                        |           |                     |                         |       |     |              |
|   |            |            |           |                        |           |                     |                         |       |     |              |

File: assets/cart.js [hunk 17]
```diff
@@ -120,8 +121,11 @@ class CartItems extends HTMLElement {
           elementToReplace.innerHTML = 
             this.getSectionInnerHTML(parsedState.sections[section.section], section.selector);
         }));
-        const updatedValue = parsedState.items[line - 1].quantity;
-        const message = updatedValue !== parseInt(quantityElement.value) ? window.cartStrings.quantityError.replace('[quantity]', updatedValue) : '';
+        const updatedValue = parsedState.items[line - 1]?.quantity;
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/cart.js:124]: I think the main reason was to address issues with a version of Safari that we officially supported at the time - https://github.com/Shopify/dawn/pull/489. But there's also a comment about code readability. Maybe we can revisit this as part of our code linting/best practices. For now, I can remove it and replace with a regular `if` statement.

File: assets/product-info.js [hunk 18]
```diff
@@ -45,57 +55,42 @@ if (!customElements.get('product-info')) {
...
-      })
-      .catch(e => {
-        console.error(e);
-      });
-    }
-
-    fetchQtyRules() {
       fetch(`${this.dataset.url}?variant=${this.currentVariant.value}&section_id=${this.dataset.section}`).then((response) => {
         return response.text()
       })
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/product-info.js:62]: Do you see any JS errors in the console? I'm able to see complimentary products as B2B user

https://user-images.githubusercontent.com/3710548/216151861-0053f075-4031-4b8c-bead-a00667ce27a5.mp4

[the remark this answers is not in the document]
Comment (sofiamatulis) [reply assets/product-info.js:62]: I think it was caching 🤔 Seems to work now 👍

File: sections/main-product.liquid [hunk 19]
```diff
@@ -308,9 +309,9 @@
...
 
             {%- when 'variant_picker' -%}
-              {% render 'product-variant-picker', block: block, product: product %}
+              {% render 'product-variant-picker', product: product, block: block, product_form_id: product_form_id %}
             {%- when 'buy_buttons' -%}
-              {%- render 'buy-buttons', block: block, product: product, product_form_id: product_form_id, show_pickup_availability: true -%}
+              {%- render 'buy-buttons', block: block, product: product, product_form_id: product_form_id, section_id: section.id, show_pickup_availability: true -%}
             {%- when 'rating' -%}
               {%- if product.metafields.reviews.rating.value != blank -%}
                 {% liquid
```
Comment (LucasLacerdaUX) [inline sections/main-product.liquid:317]: **Can be addressed in a smaller, separate PR. Not a blocker to ship this PR, possible blocker for release.**

Not related to this line, just didn't have anywhere else to leave this suggestion.

It seems like Quick Buy styles are broken for products with no image:
image

Expected:
image

It seems like the `max-width: 100%;` is taking over the `max-width: 57rem`, making the area full width and not centered.

File: sections/main-product.liquid [hunk 20]
```diff
@@ -255,7 +256,7 @@
                     {% render 'icon-plus' %}
                   </button>
                 </quantity-input>
-                <div class="quantity__rules caption">
+                <div class="quantity__rules caption no-js-hidden">
```
Comment (LucasLacerdaUX) [inline sections/main-product.liquid:259]: **Not a blocker. A11Y**
This reads as `Increments of 7 **BULLET** Minimum of 49 **BULLET** Maximum of 1183`, which IMO, isn't great.

We may be able to use empty alternative text on the `content `property by adding a `/` and setting its content to empty:
`  content: '\2022' \ '';`

Can also be addressed in a separate PR.

**Review by LucasLacerdaUX — APPROVED** [review]
Finished testing for B2B. It seems to be working as expected with no major / blocking bugs, other than possibly the Buy now button missing in this context.

We can address any issues in follow-up PRs. I've highlighted which ones are possible blockers for release, but none of them are blockers to merge this PR as they're likely not introduced by your changes.

|                                 | Product page | Featured Product @ Homepage | Featured Product @ Cart | Featured Product @ Collection + Quick Buy | Quick Buy |
|---------------------------------|--------------|-----------------------------|-------------------------|-------------------------------------------|-----------|
| [Cart Type] Drawer                  |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Cart Type] Popup                   |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Cart Type] Page                    |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Variant picker] Dropdown        |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Variant picker] Pills           |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Action] Add to Cart            |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Action] Edit quantity in Cart  |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Action] Remove from Cart       |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Product Type] with Variants          |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Products Type] no variants       |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Products Type] Sold out / unavailable |       ✅      |              ✅              |            ✅            |                     ✅                     |     ✅     |
| [Hidden block] Quantity input           |       ❌      |              ❌              |            ❌            |                     ❌                     |     ❌     |
| [Hidden block] Variant picker hidden           |       ✅`*`      |              ✅`*`              |            ✅`*`            |                     ✅`*`                     |     ✅`*`     |
| No-JS                           |       ❓      |              ❓              |            ❓            |                     ❓                     |           |

`*` = required on NoJS
❌ = quantity input is required for quantity rules. As a future enhancement, should we consider having a default `input name="quantity"` with value set to `min`?
❓ = the checkout flow isn't broken, but quantity rules are not enforced due to no-JS limitations

File: assets/global.js [hunk 21]
```diff
@@ -934,7 +935,14 @@ class VariantSelects extends HTMLElement {
 
         if (inventoryDestination) inventoryDestination.classList.toggle('visibility-hidden', inventorySource.innerText === '');
 
-        this.toggleAddButton(!this.currentVariant.available, window.variantStrings.soldOut);
+        const addButtonUpdated = html.getElementById(`ProductSubmitButton-${sectionId}`);
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js:938]: I think Buy Now buttons should work in B2B but the `spb.en.js` that's handling payments is throwing an error for B2B users. Not sure if it's a config issue in my store or a bug. Going to open an issue and check with B2B team.

Screenshot 2023-02-01 at 4 18 56 PM

File: assets/global.js [hunk 22]
```diff
@@ -934,7 +935,14 @@ class VariantSelects extends HTMLElement {
...
+        this.toggleAddButton(addButtonUpdated ? addButtonUpdated.hasAttribute('disabled') : true, window.variantStrings.soldOut);
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js:939]: Due to the minimum quantity rule being higher than available inventory, we disabled the add-to-cart button. But you're right that we probably also want to show sold out badge and add sold out styling to variant picker for consistency.

File: sections/main-product.liquid (hunk 20 above)
Comment (KaichenWang) [reply sections/main-product.liquid:259]: Issue created - https://github.com/Shopify/dawn/issues/2275

File: sections/main-product.liquid (hunk 19 above)
Comment (KaichenWang) [reply sections/main-product.liquid:317]: Issue created - https://github.com/Shopify/dawn/issues/2274

File: assets/global.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/global.js:938]: Issue created - https://github.com/Shopify/dawn/issues/2273
